lms/services/lti_outcomes.py

```diff
@@ -1,26 +1,74 @@
+from datetime import datetime, timezone
 from xml.parsers.expat import ExpatError
 
 import xmltodict
 
+from lms.services import LTIAHTTPService
 from lms.services.exceptions import ExternalRequestError
 
 __all__ = ["LTIOutcomesClient"]
 
 
 class LTIOutcomesClient:
     """
-    Service for making requests to an LMS's Outcomes Management endpoint.
+    Service for sending grades back to the LMS.
 
-    See https://www.imsglobal.org/specs/ltiomv1p0/specification.
+    See:
+         LTI1.1 Outcomes https://www.imsglobal.org/specs/ltiomv1p0/specification.
+         LTI1.3 Assignment and Grade Services https://www.imsglobal.org/spec/lti-ags/v2p0
     """
 
-    def __init__(self, _context, request):
-        self.oauth1_service = request.find_service(name="oauth1")
-        self.http_service = request.find_service(name="http")
+    LTIA_SCOPES = [
+        "https://purl.imsglobal.org/spec/lti-ags/scope/lineitem",
+        "https://purl.imsglobal.org/spec/lti-ags/scope/result.readonly",
+        "https://purl.imsglobal.org/spec/lti-ags/scope/score",
+    ]
 
-        self.service_url = request.parsed_params["lis_outcome_service_url"]
+    def __init__(  # pylint:disable=too-many-arguments
+        self, oauth1_service, http_service, ltia_service, grading_url, lti_version
+    ):
+        self.oauth1_service = oauth1_service
+        self.http_service = http_service
+        self.ltia_service = ltia_service
 
-    def read_result(self, lis_result_sourcedid):
+        self.service_url = grading_url
+        self.lti_version = lti_version
+
+    def read_result(self, lis_result_sourcedid=None):
+        if self.lti_version == "1.3.0":
+            return self._read_result_lti_v13(lis_result_sourcedid)
+
+        return self._read_result_lti_v11(lis_result_sourcedid)
+
+    def record_result(self, lis_result_sourcedid, score=None, pre_record_hook=None):
+        if self.lti_version == "1.3.0":
+            return self._record_result_lti_v13(lis_result_sourcedid, score=score)
+
+        return self._record_result_lti_v11(
+            lis_result_sourcedid, score=score, pre_record_hook=pre_record_hook
+        )
+
+    def _read_result_lti_v13(self, lis_result_sourcedid):
+        try:
+            response = self.ltia_service.request(
+                self.LTIA_SCOPES,
+                "GET",
+                self.service_url + "/results",
+                params={"user_id": lis_result_sourcedid},
+                headers={"Accept": "application/vnd.ims.lis.v2.resultcontainer+json"},
+            )
+        except ExternalRequestError as err:
+            if err.status_code == 404:
+                return None
+            raise
+
+        results = response.json()
+        if not results:
+            return None
+
+        return results[-1]["resultScore"] / results[-1]["resultMaximum"]
+
```

Comment:
This looks very easy to make explode? In the other class there's some try catching around to prevent this from going mad. Also fun opportunity for a divide by zero here, which you don't see every day.

Reply:
Not sure about catching KeyError and friends. I think if some LMS breaks the spec we rather find out in sentry/logs before. If it does happen often and we have to handle we'll see.

I see division by zero more plausible. But not sure what to do, return None? return the absolute `resultScore`?